### PR TITLE
feat: redefine primary keys

### PR DESCRIPTION
**Pull request by purplesmoke05**
Close: #1377

### Changes

- redefine table schemas so that they have preferable primary keys.
  - [x] migration for `position`
  - [x] migration for `executable_contract`
  - [x] migration for `idx_position_bond_block_number`
  - [x] migration for `idx_position_share_block_number`
  - [x] migration for `idx_position_coupon_block_number`
  - [x] migration for `idx_position_membership_block_number`

- Fixed some **GROUP BY** queries in APIs which contains dropped columns in this PR.
  - `app/api/routers/position.py`
  - `app/api/routers/token.py`

- Added migration **downgrade** test.

### Migration Results

<details>
<summary>Postgres</summary>

```bash
$ DATABASE_URL=postgresql+psycopg://apluser:apluserpass@localhost:5432/apldb poetry run ./run_migration.sh init
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> a80595c53d52, 026_v22.3
INFO  [alembic.runtime.migration] Running upgrade a80595c53d52 -> e8d970fdd886, 034_v22.6
INFO  [alembic.runtime.migration] Running upgrade e8d970fdd886 -> 55dee53de2ca, 041_v22.9
INFO  [alembic.runtime.migration] Running upgrade 55dee53de2ca -> 446f913d1f41, 047_v22.12
INFO  [alembic.runtime.migration] Running upgrade 446f913d1f41 -> 1055cb068506, 054_v23.3
INFO  [alembic.runtime.migration] Running upgrade 1055cb068506 -> ca8ff8c0c01d, v23.6_feature/alembic
INFO  [alembic.runtime.migration] Running upgrade ca8ff8c0c01d -> 40ee030e9302, v23_6_feature_1377
$ DATABASE_URL=postgresql+psycopg://apluser:apluserpass@localhost:5432/apldb poetry run ./run_migration.sh downgrade
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade 40ee030e9302 -> ca8ff8c0c01d, v23_6_feature_1377
INFO  [alembic.runtime.migration] Running downgrade ca8ff8c0c01d -> 1055cb068506, v23.6_feature/alembic
INFO  [alembic.runtime.migration] Running downgrade 1055cb068506 -> 446f913d1f41, 054_v23.3
INFO  [alembic.runtime.migration] Running downgrade 446f913d1f41 -> 55dee53de2ca, 047_v22.12
INFO  [alembic.runtime.migration] Running downgrade 55dee53de2ca -> e8d970fdd886, 041_v22.9
INFO  [alembic.runtime.migration] Running downgrade e8d970fdd886 -> a80595c53d52, 034_v22.6
INFO  [alembic.runtime.migration] Running downgrade a80595c53d52 -> , 026_v22.3
$ DATABASE_URL=postgresql+psycopg://apluser:apluserpass@localhost:5432/apldb poetry run ./run_migration.sh upgrade
./run_migration.sh: line 20: /Users/y-otosu/.bash_profile: No such file or directory
./run_migration.sh: line 21: cd: /app/ibet-Wallet-API: No such file or directory
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> a80595c53d52, 026_v22.3
INFO  [alembic.runtime.migration] Running upgrade a80595c53d52 -> e8d970fdd886, 034_v22.6
INFO  [alembic.runtime.migration] Running upgrade e8d970fdd886 -> 55dee53de2ca, 041_v22.9
INFO  [alembic.runtime.migration] Running upgrade 55dee53de2ca -> 446f913d1f41, 047_v22.12
INFO  [alembic.runtime.migration] Running upgrade 446f913d1f41 -> 1055cb068506, 054_v23.3
INFO  [alembic.runtime.migration] Running upgrade 1055cb068506 -> ca8ff8c0c01d, v23.6_feature/alembic
INFO  [alembic.runtime.migration] Running upgrade ca8ff8c0c01d -> 40ee030e9302, v23_6_feature_1377
```
</details>

<details>
<summary>MySQL</summary>

```bash
$ DATABASE_URL=mysql+pymysql://apluser:apluserpass@localhost:3306/apldb poetry run ./run_migration.sh init
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> a80595c53d52, 026_v22.3
INFO  [alembic.runtime.migration] Running upgrade a80595c53d52 -> e8d970fdd886, 034_v22.6
INFO  [alembic.runtime.migration] Running upgrade e8d970fdd886 -> 55dee53de2ca, 041_v22.9
INFO  [alembic.runtime.migration] Running upgrade 55dee53de2ca -> 446f913d1f41, 047_v22.12
INFO  [alembic.runtime.migration] Running upgrade 446f913d1f41 -> 1055cb068506, 054_v23.3
INFO  [alembic.runtime.migration] Running upgrade 1055cb068506 -> ca8ff8c0c01d, v23.6_feature/alembic
INFO  [alembic.runtime.migration] Running upgrade ca8ff8c0c01d -> 40ee030e9302, v23_6_feature_1377
$ DATABASE_URL=mysql+pymysql://apluser:apluserpass@localhost:3306/apldb poetry run ./run_migration.sh downgrade
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade 40ee030e9302 -> ca8ff8c0c01d, v23_6_feature_1377
INFO  [alembic.runtime.migration] Running downgrade ca8ff8c0c01d -> 1055cb068506, v23.6_feature/alembic
INFO  [alembic.runtime.migration] Running downgrade 1055cb068506 -> 446f913d1f41, 054_v23.3
INFO  [alembic.runtime.migration] Running downgrade 446f913d1f41 -> 55dee53de2ca, 047_v22.12
INFO  [alembic.runtime.migration] Running downgrade 55dee53de2ca -> e8d970fdd886, 041_v22.9
INFO  [alembic.runtime.migration] Running downgrade e8d970fdd886 -> a80595c53d52, 034_v22.6
INFO  [alembic.runtime.migration] Running downgrade a80595c53d52 -> , 026_v22.3
$ DATABASE_URL=mysql+pymysql://apluser:apluserpass@localhost:3306/apldb poetry run ./run_migration.sh upgrade
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> a80595c53d52, 026_v22.3
INFO  [alembic.runtime.migration] Running upgrade a80595c53d52 -> e8d970fdd886, 034_v22.6
INFO  [alembic.runtime.migration] Running upgrade e8d970fdd886 -> 55dee53de2ca, 041_v22.9
INFO  [alembic.runtime.migration] Running upgrade 55dee53de2ca -> 446f913d1f41, 047_v22.12
INFO  [alembic.runtime.migration] Running upgrade 446f913d1f41 -> 1055cb068506, 054_v23.3
INFO  [alembic.runtime.migration] Running upgrade 1055cb068506 -> ca8ff8c0c01d, v23.6_feature/alembic
INFO  [alembic.runtime.migration] Running upgrade ca8ff8c0c01d -> 40ee030e9302, v23_6_feature_1377
```
</details>

### CODE REVIEW

File: app/api/routers/position.py [hunk 1]
```diff
@@ -525,7 +525,8 @@ def get_list_from_index(
             )
             .group_by(
                 Listing.id,
-                IDXPosition.id,
+                IDXPosition.token_address,
+                IDXPosition.account_address,
```
Comment (purplesmoke05) [inline app/api/routers/position.py:529]: Fixed **GROUP BY** key

File: app/api/routers/token.py [hunk 2]
```diff
@@ -187,13 +187,17 @@ def get_token_holders(
...
+            IDXPosition.account_address,
+            IDXLockedPosition.account_address,
+        )
     )
     if request_query.exclude_owner is True:
         query = query.filter(IDXPosition.account_address != listed_token.owner_address)
 
     holders: list[tuple[IDXPosition, int | None]] = query.order_by(
-        desc(IDXPosition.id)
+        desc(IDXPosition.created)
```
Comment (purplesmoke05) [inline app/api/routers/token.py:200]: Replaced `id` with `created` because `position.id` is dropped.

File: app/api/routers/token.py [hunk 3]
```diff
@@ -266,12 +270,16 @@ def get_token_holders_count(
...
+            IDXPosition.token_address,
+            IDXPosition.account_address,
+            IDXLockedPosition.account_address,
+        )
     )
     if request_query.exclude_owner is True:
         query = query.filter(IDXPosition.account_address != listed_token.owner_address)
 
-    _count = query.order_by(desc(IDXPosition.id)).count()
+    _count = query.count()
```
Comment (purplesmoke05) [inline app/api/routers/token.py:282]: Removed unnecessary **ORDER BY** clause.

File: migrations/versions/40ee030e9302_v23_6_feature_1377.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+    op.get_bind().execute(sa.text(f"DELETE FROM position WHERE modified IS null;"))
+    op.get_bind().execute(
+        sa.text(
+            f"""DELETE FROM position WHERE (modified) NOT IN (
+                    SELECT latest_position.max_modified AS modified FROM (
+                        SELECT  max(modified) AS max_modified FROM position GROUP BY token_address, account_address
+                    ) AS latest_position
+                );"""
+        )
+    )
```
Comment (purplesmoke05) [inline migrations/versions/40ee030e9302_v23_6_feature_1377.py:42]: - To remove duplicate rows, added SQL statement to erase all but the most recent data. 
- Whether the data is up-to-date or not is determined by **modified** field.
- Such SQL is executed against the following tables.
  - `position`
  - `executable_contract`
  - `idx_position_bond_block_number`
  - `idx_position_share_block_number `
  - `idx_position_coupon_block_number `
  - `idx_position_membership_block_number `

File: tests/migrations/upgrade_test.py [hunk 5]
```diff
@@ -303,12 +339,398 @@ def test_upgrade_v23_6_to(
...
+                    "SELECT COUNT(*) FROM position WHERE token_address = 'token_address1'"
+                )
+            ).scalar()
+
+            # Ensure that there are 3 records.
+            #    (token_address1, account_address1)
+            #    (token_address1, account_address2)
+            #    (token_address2, account_address2)
+            assert all_row_count == 3
+            assert token_address_1_count == 2
```
Comment (purplesmoke05) [inline tests/migrations/upgrade_test.py:652]: - Reproduce datas that may already exists and assert the results as correct after migration.

File: tests/migrations/downgrade_test.py [hunk 6]
```diff
@@ -132,3 +169,283 @@ def get_migrate_version(db_engine: Engine):
             return connection.execute(
                 text("SELECT version FROM migrate_version")
             ).scalar()
+
+    def test_downgrade_from_v23_6(
+        self, alembic_runner: MigrationContext, caplog: LogCaptureFixture
+    ):
```
Comment (purplesmoke05) [inline tests/migrations/downgrade_test.py:175]: - Added unit tests for extra caution when executing the downgrade migration script from version 23.6.

File: tests/migrations/downgrade_test.py [hunk 7]
```diff
@@ -132,3 +169,283 @@ def get_migrate_version(db_engine: Engine):
...
+            assert all_row_count == 3
+            assert token_address_1_count == 2
+
+        for assert_table_name in table_name_list:
+            table_bf = self.create_sorted_table(tables_bf[assert_table_name])
+            table_af = self.create_sorted_table(tables_af[assert_table_name])
+            assert (
+                CreateTable(table_af).compile(engine).string
+                == CreateTable(table_bf).compile(engine).string
+            )
```
Comment (purplesmoke05) [inline tests/migrations/downgrade_test.py:451]: - Added code to verify that the following two types of databases are identical:
  - 1.database that has been upgraded to version **23.3 after initialization**
  - 2.database that has been upgraded to version 23.6 and then **downgraded to version 23.3**

**Review by YoshihitoAso — APPROVED** [review]

**Pull request merged**